File: nodejs/OnetWebService.js
"use strict";

class OnetWebService {
  constructor(api_key) {
    this._config = {
      api_key: api_key
    };
    this.set_version();
  }

  set_version(version) {
    if (version === undefined) {
      this._config.baseURL = 'https://api-v2.onetcenter.org/';
    } else {
      this._config.baseURL = 'https://api-v' + version + '.onetcenter.org/';
    }
  }

  _encode_query(query) {
    let params = new URLSearchParams();
    for (const key of Object.keys(query)) {
      const val = query[key];
      if (Array.isArray(val)) {
        for (const v of val) {
          params.append(key, v);
        }
      } else {
        params.append(key, val);
      }
    }
    return params.toString();
  }

  call(path, query) {
    let url = this._config.baseURL + path;
    if (query !== null && query !== undefined) {
      url += '?' + this._encode_query(query);
    }
    return new Promise((resolve, reject) => {
      fetch(url, {
        method: 'GET',
        headers: {
          'X-API-Key': this._config.api_key,
          'User-Agent': 'nodejs-OnetWebService/2.10 (bot)'
        }
      })
      .then((response) => {
        if (response.status == 200 || response.status == 422) {
          response.json()
            .then((data) => Object.hasOwn(data, 'error') ? reject(data.error) : resolve(data))
            .catch((error) => reject(`Call to ${url} failed on JSON parse`));
        } else {
          reject(`Call to ${url} failed with error code ${response.status}`);
        }
      })
      .catch((error) => {
        if (error.message) {
          reject(`Call to ${url} failed with reason: ${error.message}`);
        } else {
          reject(`Call to ${url} failed with unknown reason`);
        }
      });
    });
  }
}

module.exports = OnetWebService

File: csharp-dotnet/OnetWebService.cs
﻿using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json.Nodes;

public class OnetWebService
{
    public class QueryParams : IEnumerable<KeyValuePair<string, string>>
    {
        private List<KeyValuePair<string, string>> internalParams = new List<KeyValuePair<string, string>>();
        public IEnumerator<KeyValuePair<string, string>> GetEnumerator() => internalParams.GetEnumerator();
        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => internalParams.GetEnumerator();

        public void Add(string key, string value) => internalParams.Add(new KeyValuePair<string, string>(key, value));
    }

    private string baseURL;
    private HttpClient client;

    private static string EncodeAuth(string username, string password)
    {
        return Convert.ToBase64String(System.Text.Encoding.ASCII.GetBytes(username + ":" + password));
    }

    public OnetWebService(string api_key)
    {
        HttpClientHandler handler = new HttpClientHandler()
        {
            AllowAutoRedirect = false,
        };
        client = new HttpClient(handler);
        client.Timeout = new TimeSpan(0, 0, 10);
        client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        client.DefaultRequestHeaders.UserAgent.ParseAdd("dotnet-OnetWebService/2.00 (bot)");
        client.DefaultRequestHeaders.Add("X-API-Key", api_key);

        SetVersion();
    }

    public void SetVersion(string version = "")
    {
        if (version == "")
        {
            baseURL = "https://api-v2.onetcenter.org/";
        }
        else
        {
            baseURL = "https://api-v" + version + ".onetcenter.org/";
        }
    }

    public async Task<JsonNode> Call(string path, QueryParams query = null)
    {
        List<string> encoded_params = new List<string>();
        if (query != null)
        {
            foreach (KeyValuePair<string, string> pair in query)
            {
                encoded_params.Add(System.Net.WebUtility.UrlEncode(pair.Key) + "=" + System.Net.WebUtility.UrlEncode(pair.Value));
            }
        }
        string url = baseURL + path;
        if (encoded_params.Count > 0)
        {
            url += "?" + String.Join("&", encoded_params.ToArray());
        }

        JsonNode result = new JsonObject
        {
            ["error"] = "Call to " + url + " failed with unknown reason"
        };

        try
        {
            HttpResponseMessage response = await client.GetAsync(url);
            if (response.StatusCode == (System.Net.HttpStatusCode)200 || response.StatusCode == (System.Net.HttpStatusCode)422)
            {
                result = JsonNode.Parse(await response.Content.ReadAsStringAsync());
            }
            else
            {
                result["error"] = "Call to " + url + " failed with error code " + ((int)response.StatusCode).ToString();
            }
        }
        catch (HttpRequestException e)
        {
            result["error"] = "Call to " + url + " failed with reason: " + e.Message;
        }

        return result;
    }
}
